mitiq/pec/representations/learning.py

```diff
@@ -22,11 +22,112 @@
+) -> Tuple[bool, float, float]:
+    r"""This function learns the depolarizing noise parameter (epsilon)
+    associated to a set of input operations. The learning process is based on
+    the execution of a set of training circuits on a noisy backend and on a
+    classical simulator. The training circuits are near-Clifford approximations
+    of the input circuit. A depolarizing noise model characterization is
+    assumed.
+
+    Args:
+        operations_to_learn: The ideal operations to learn the noise model of.
+        circuit: The full quantum program as defined by the user.
+        ideal_executor: Simulates a circuit and returns
+            a noiseless ``QuantumResult``.
+        noisy_executor: Executes a circuit on a noisy backend
+            and returns a ``QuantumResult``.
+        pec_kwargs: Options to pass to ``execute_w_pec`` for the
+            error-mitigated expectation value obtained from executing
+            the training circuits.
+        num_training_circuits: Number of near-Clifford circuits to be
+            generated for training.
+        fraction_non_clifford: The (approximate) fraction of non-Clifford
+            gates in each training circuit.
+        training_random_state: Seed for sampling the training circuits.
+        epsilon0: Initial guess for noise strength.
+        eta0: Initial guess for noise bias.
+        observable (optional): Observable to compute the expectation value of.
+            If None, the ``executor`` must return an expectation value.
+            Otherwise the `QuantumResult` returned by `executor` is used to
+            compute the expectation of the observable.
+        learning_kwargs (optional): Additional data and options including
+            ``pec_data`` from pre-executed runs of PEC on training circuits,
+            ``method`` an optimization method supported by
+            ``scipy.optimize.minimize``, and settings for the chosen
+            optimization method.
+
+    Returns:
+        A flag indicating whether or not the optimizer exited successfully, the
+        optimized noise strength epsilon, and the optimized noise bias, eta.
+
+    .. note:: Using this function may require some tuning. One of the main
+        challenges is setting a good value of ``num_samples`` in the PEC
+        options ``pec_kwargs``. Setting a small value of ``num_samples`` is
+        typically necessary to obtain a reasonable execution time. On the other
+        hand, using a number of PEC samples that is too small can result in a
+        large statistical error, ultimately causing the optimization process to
+        fail.
+    """
+    training_circuits = generate_training_circuits(
+        circuit=circuit,
+        num_training_circuits=num_training_circuits,
+        fraction_non_clifford=fraction_non_clifford,
+        random_state=training_random_state,
+    )
+
+    ideal_values = np.array(
+        [ideal_executor.evaluate(t, observable) for t in training_circuits]
+    )
```

Comment:
Both good questions!
1. I'm not entirely sure about the numpy array, need to trace back to where that came from.
2. I've kept the executors separate because someday (hopefully!) someone will run the noisy executor on a noisy backend and the ideal executor on a classical computer.

Reply:
Update: `ideal_values` doesn't have to be a numpy array. Is there a type you think would be preferable over a numpy array?